# Deep Learning Lab 3

* Mount the drive and load your data.
* Create a Neural network to run on your dataset.
* Tune the model for best accuracy.
* Evaluate and show the results of the model for the dataset.

In [1]:
import numpy as np
import pandas as pd

## Loading Datasets

### Mount drive and load local data

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
df = pd.read_csv('/content/drive/MyDrive/Deep Learning/Datasets/Preprocessed Audio MFCC.csv')
df.head()

0           1          2  ...        28        29  class
0 -507.188232  115.508881  -0.499709  ... -4.737616 -5.110260     आम
1 -504.667755   67.896660  -8.077479  ...  1.422675 -0.993730     आम
2 -374.015472   69.507240 -44.598995  ... -4.217253  0.798838     आम
3 -567.320007   74.504440   3.016104  ... -1.867280 -1.706182     आम
4 -578.482422   86.136101   4.080464  ... -2.596040 -2.900922     आम

[5 rows x 31 columns]

The dataset above was extracted by preprocessing audio data which we accumulated in the 5th semester to work on a speech recognition project. It contains 30 Mel Frequency Cepstral Coefficient Means useful for predicting 10 words.

In [4]:
df.shape

(1046, 31)

### Splitting data into train, validate and test

In [5]:
from sklearn.model_selection import train_test_split as tts
from sklearn.preprocessing import MinMaxScaler, StandardScaler

In [6]:
ss = StandardScaler()
x = ss.fit_transform(df.drop(columns=['class']))
x

array([[-0.90278576,  0.70030512,  0.66706866, ...,  0.04617815,
        -1.01955239, -1.30886155],
       [-0.87773383, -0.81638491,  0.22861753, ..., -0.15684798,
         0.70715394, -0.09567425],
       [ 0.42086645, -0.76507978, -1.88452384, ...,  0.03493916,
        -0.87369663,  0.43261585],
       ...,
       [ 1.43046754,  0.39350795, -1.88564223, ...,  3.55744374,
         0.83796201,  0.54367184],
       [ 1.1323243 ,  0.61046706, -0.7691001 , ..., -0.68305318,
        -1.30435711, -0.97295748],
       [-1.17867585, -0.05619563,  0.16408634, ...,  0.8599921 ,
         0.20971642,  1.33654238]])

In [7]:
y = df['class']
num_labels = len(y.unique())
y = np.array(pd.get_dummies(y))

In [8]:
num_labels

10

In [9]:
xt, xts, yt, yts = tts(x, y, test_size=0.1, stratify=y)

In [10]:
xtr, xv, ytr, yv = tts(xt, yt, test_size=0.1, stratify=yt)

In [11]:
xtr.shape

(846, 30)

## Model Creation

In [12]:
import tensorflow as tf

In [13]:
lr=0.01

model = tf.keras.Sequential()
model.add(tf.keras.layers.Dense(xtr.shape[1], input_shape=(xtr.shape[1],), activation='relu'))
model.add(tf.keras.layers.Dense(num_labels, activation='softmax'))

loss_fn = tf.keras.losses.CategoricalCrossentropy()
opt = tf.keras.optimizers.Adam(learning_rate=lr)

model.compile(optimizer=opt, loss=loss_fn, metrics=['accuracy'])
model.fit(xtr, ytr, epochs=30, validation_data=(xv, yv))

Epoch 1/30
27/27 [==============================] - 1s 8ms/step - loss: 2.3693 - accuracy: 0.1300 - val_loss: 2.2981 - val_accuracy: 0.1684
Epoch 2/30
27/27 [==============================] - 0s 2ms/step - loss: 2.0922 - accuracy: 0.2435 - val_loss: 2.2369 - val_accuracy: 0.2105
Epoch 3/30
27/27 [==============================] - 0s 2ms/step - loss: 1.9548 - accuracy: 0.3085 - val_loss: 2.2766 - val_accuracy: 0.2526
Epoch 4/30
27/27 [==============================] - 0s 2ms/step - loss: 1.8517 - accuracy: 0.3534 - val_loss: 2.2294 - val_accuracy: 0.2632
Epoch 5/30
27/27 [==============================] - 0s 2ms/step - loss: 1.7736 - accuracy: 0.3771 - val_loss: 2.2858 - val_accuracy: 0.2632
Epoch 6/30
27/27 [==============================] - 0s 2ms/step - loss: 1.7389 - accuracy: 0.3712 - val_loss: 2.2817 - val_accuracy: 0.2211
Epoch 7/30
27/27 [==============================] - 0s 2ms/step - loss: 1.6738 - accuracy: 0.4019 - val_loss: 2.2559 - val_accuracy: 0.2632
Epoch 8/30
27/27 [==

## Hyperparameter Tuning

### Changing the Activation function

In [14]:
model = tf.keras.Sequential()
model.add(tf.keras.layers.Dense(xtr.shape[1], input_shape=(xtr.shape[1],), activation='tanh'))
model.add(tf.keras.layers.Dense(num_labels, activation='softmax'))

model.compile(optimizer=opt, loss=loss_fn, metrics=['accuracy'])
model.fit(xtr, ytr, epochs=30, validation_data=(xv, yv))

Epoch 1/30
27/27 [==============================] - 1s 7ms/step - loss: 2.3929 - accuracy: 0.1702 - val_loss: 2.3597 - val_accuracy: 0.2105
Epoch 2/30
27/27 [==============================] - 0s 2ms/step - loss: 2.0624 - accuracy: 0.2624 - val_loss: 2.2418 - val_accuracy: 0.2632
Epoch 3/30
27/27 [==============================] - 0s 2ms/step - loss: 1.8655 - accuracy: 0.3404 - val_loss: 2.2971 - val_accuracy: 0.2421
Epoch 4/30
27/27 [==============================] - 0s 2ms/step - loss: 1.7712 - accuracy: 0.3546 - val_loss: 2.2618 - val_accuracy: 0.2421
Epoch 5/30
27/27 [==============================] - 0s 2ms/step - loss: 1.6674 - accuracy: 0.4149 - val_loss: 2.3690 - val_accuracy: 0.2842
Epoch 6/30
27/27 [==============================] - 0s 2ms/step - loss: 1.5862 - accuracy: 0.4243 - val_loss: 2.4038 - val_accuracy: 0.2211
Epoch 7/30
27/27 [==============================] - 0s 2ms/step - loss: 1.5429 - accuracy: 0.4409 - val_loss: 2.4068 - val_accuracy: 0.2737
Epoch 8/30
27/27 [==

The accuracy with 'tanh' is not too different from the one with 'relu'. Since 'relu' is a simpler activation function, we choose to move forward with it.

### Adding a Dropout layer

In [15]:
model = tf.keras.Sequential()
model.add(tf.keras.layers.Dense(xtr.shape[1], input_shape=(xtr.shape[1],), activation='relu'))
model.add(tf.keras.layers.Dropout(0.4))
model.add(tf.keras.layers.Dense(num_labels, activation='softmax'))

model.compile(optimizer=opt, loss=loss_fn, metrics=['accuracy'])
model.fit(xtr, ytr, epochs=30, validation_data=(xv, yv))

Epoch 1/30
27/27 [==============================] - 1s 7ms/step - loss: 2.5806 - accuracy: 0.1312 - val_loss: 2.1822 - val_accuracy: 0.2421
Epoch 2/30
27/27 [==============================] - 0s 2ms/step - loss: 2.2449 - accuracy: 0.2080 - val_loss: 2.1242 - val_accuracy: 0.1579
Epoch 3/30
27/27 [==============================] - 0s 2ms/step - loss: 2.0800 - accuracy: 0.2447 - val_loss: 2.1746 - val_accuracy: 0.2947
Epoch 4/30
27/27 [==============================] - 0s 2ms/step - loss: 2.0321 - accuracy: 0.2636 - val_loss: 2.1364 - val_accuracy: 0.2842
Epoch 5/30
27/27 [==============================] - 0s 2ms/step - loss: 1.9731 - accuracy: 0.2908 - val_loss: 2.1511 - val_accuracy: 0.2737
Epoch 6/30
27/27 [==============================] - 0s 3ms/step - loss: 1.9402 - accuracy: 0.2908 - val_loss: 2.1249 - val_accuracy: 0.2947
Epoch 7/30
27/27 [==============================] - 0s 2ms/step - loss: 1.9091 - accuracy: 0.3156 - val_loss: 2.1610 - val_accuracy: 0.2632
Epoch 8/30
27/27 [==

Adding a dropout layer has reduced the overfitting but it hasn't helped with the accuracy. Let us try some other methods as well.

### Applying Regularisation

#### L1 Regularisation

In [16]:
model = tf.keras.Sequential()
model.add(tf.keras.layers.Dense(xtr.shape[1], input_shape=(xtr.shape[1],), activation='relu', kernel_regularizer=tf.keras.regularizers.L1(0.02)))
model.add(tf.keras.layers.Dense(num_labels, activation='softmax'))

model.compile(optimizer=opt, loss=loss_fn, metrics=['accuracy'])
model.fit(xtr, ytr, epochs=30, validation_data=(xv, yv))

Epoch 1/30
27/27 [==============================] - 1s 8ms/step - loss: 3.5390 - accuracy: 0.1182 - val_loss: 2.9310 - val_accuracy: 0.1158
Epoch 2/30
27/27 [==============================] - 0s 2ms/step - loss: 2.6613 - accuracy: 0.1785 - val_loss: 2.4751 - val_accuracy: 0.2000
Epoch 3/30
27/27 [==============================] - 0s 2ms/step - loss: 2.4033 - accuracy: 0.2199 - val_loss: 2.3551 - val_accuracy: 0.2632
Epoch 4/30
27/27 [==============================] - 0s 2ms/step - loss: 2.3285 - accuracy: 0.2435 - val_loss: 2.2835 - val_accuracy: 0.2526
Epoch 5/30
27/27 [==============================] - 0s 2ms/step - loss: 2.3033 - accuracy: 0.2600 - val_loss: 2.2351 - val_accuracy: 0.2211
Epoch 6/30
27/27 [==============================] - 0s 2ms/step - loss: 2.2725 - accuracy: 0.2329 - val_loss: 2.3145 - val_accuracy: 0.1684
Epoch 7/30
27/27 [==============================] - 0s 2ms/step - loss: 2.2266 - accuracy: 0.2660 - val_loss: 2.3448 - val_accuracy: 0.2316
Epoch 8/30
27/27 [==

L1 regularisation seems to have worked better than Dropout in reducing overfitting. But like Dropout, it doesn't help imporve the accuracy much.

#### L2 Regularisation

In [17]:
model = tf.keras.Sequential()
model.add(tf.keras.layers.Dense(xtr.shape[1], input_shape=(xtr.shape[1],), activation='relu', kernel_regularizer=tf.keras.regularizers.L2(0.3)))
model.add(tf.keras.layers.Dense(num_labels, activation='softmax'))

model.compile(optimizer=opt, loss=loss_fn, metrics=['accuracy'])
model.fit(xtr, ytr, epochs=30, validation_data=(xv, yv))

Epoch 1/30
27/27 [==============================] - 1s 7ms/step - loss: 3.9682 - accuracy: 0.1040 - val_loss: 2.7136 - val_accuracy: 0.1158
Epoch 2/30
27/27 [==============================] - 0s 2ms/step - loss: 2.4886 - accuracy: 0.1182 - val_loss: 2.3921 - val_accuracy: 0.1053
Epoch 3/30
27/27 [==============================] - 0s 2ms/step - loss: 2.3617 - accuracy: 0.1537 - val_loss: 2.2904 - val_accuracy: 0.1895
Epoch 4/30
27/27 [==============================] - 0s 2ms/step - loss: 2.3542 - accuracy: 0.1442 - val_loss: 2.3198 - val_accuracy: 0.1789
Epoch 5/30
27/27 [==============================] - 0s 2ms/step - loss: 2.3360 - accuracy: 0.1596 - val_loss: 2.3390 - val_accuracy: 0.2000
Epoch 6/30
27/27 [==============================] - 0s 2ms/step - loss: 2.3170 - accuracy: 0.1525 - val_loss: 2.3071 - val_accuracy: 0.1263
Epoch 7/30
27/27 [==============================] - 0s 2ms/step - loss: 2.2927 - accuracy: 0.1785 - val_loss: 2.2668 - val_accuracy: 0.1684
Epoch 8/30
27/27 [==

L2 regularisation seems to show a similar effect as L1. 

#### Using all of them together

In [18]:
model = tf.keras.Sequential()
model.add(tf.keras.layers.Dense(xtr.shape[1], input_shape=(xtr.shape[1],), activation='relu', kernel_regularizer=tf.keras.regularizers.L1L2(0.01, 0.05)))
model.add(tf.keras.layers.Dropout(0.2))
model.add(tf.keras.layers.Dense(num_labels, activation='softmax'))

model.compile(optimizer=opt, loss=loss_fn, metrics=['accuracy'])
model.fit(xtr, ytr, epochs=30, validation_data=(xv, yv))

Epoch 1/30
27/27 [==============================] - 1s 8ms/step - loss: 3.3624 - accuracy: 0.1111 - val_loss: 2.7070 - val_accuracy: 0.1684
Epoch 2/30
27/27 [==============================] - 0s 2ms/step - loss: 2.5364 - accuracy: 0.1229 - val_loss: 2.3697 - val_accuracy: 0.1158
Epoch 3/30
27/27 [==============================] - 0s 2ms/step - loss: 2.3535 - accuracy: 0.1288 - val_loss: 2.3096 - val_accuracy: 0.2000
Epoch 4/30
27/27 [==============================] - 0s 2ms/step - loss: 2.3132 - accuracy: 0.1418 - val_loss: 2.2747 - val_accuracy: 0.1579
Epoch 5/30
27/27 [==============================] - 0s 2ms/step - loss: 2.2937 - accuracy: 0.1584 - val_loss: 2.2785 - val_accuracy: 0.1684
Epoch 6/30
27/27 [==============================] - 0s 2ms/step - loss: 2.3143 - accuracy: 0.1478 - val_loss: 2.2632 - val_accuracy: 0.1474
Epoch 7/30
27/27 [==============================] - 0s 2ms/step - loss: 2.2976 - accuracy: 0.1418 - val_loss: 2.2631 - val_accuracy: 0.1789
Epoch 8/30
27/27 [==

### Varying number of layers and nodes

In [19]:
def getAccuracy(numoflayers, numofnodes):
    
    # Initialise Neural network
    model=tf.keras.models.Sequential()
    model.add(tf.keras.layers.Dense(xtr.shape[1], input_shape=(xtr.shape[1],), activation='relu'))
    for i in range(1, numoflayers+1):
        model.add(tf.keras.layers.Dense(xtr.shape[1]*numofnodes, activation='relu'))
        model.add(tf.keras.layers.Dropout(0.2))
    model.add(tf.keras.layers.Dense(num_labels, activation='softmax'))

    # Decide parameters
    model.compile(optimizer=opt, loss=loss_fn, metrics=['accuracy'])
    
    # Train model and get accuracy
    model.fit(xtr, ytr, epochs=30, verbose=0)
    val_accuracy=model.evaluate(xv, yv, verbose=0)[1]
    
    return val_accuracy

In [20]:
from datetime import datetime
start = datetime.now()
tries = 5
for layer in range(1, 4):
    for node in range(1,11):
        acc = 0
        for i in range(tries):
            acc += getAccuracy(layer, node)
        acc /= tries
        print(f'Layers - {layer}\tNodes - {node}\tAccuracy - {round(acc*100, 5)}')
duration = datetime.now() - start
print("Training completed in time: ", duration)

Layers - 1	Nodes - 1	Accuracy - 18.73684
Layers - 1	Nodes - 2	Accuracy - 22.73684
Layers - 1	Nodes - 3	Accuracy - 22.52632
Layers - 1	Nodes - 4	Accuracy - 22.94737
Layers - 1	Nodes - 5	Accuracy - 23.57895
Layers - 1	Nodes - 6	Accuracy - 23.78947
Layers - 1	Nodes - 7	Accuracy - 23.57895
Layers - 1	Nodes - 8	Accuracy - 20.21053
Layers - 1	Nodes - 9	Accuracy - 22.73684
Layers - 1	Nodes - 10	Accuracy - 21.89474
Layers - 2	Nodes - 1	Accuracy - 14.10526
Layers - 2	Nodes - 2	Accuracy - 16.63158
Layers - 2	Nodes - 3	Accuracy - 17.05263
Layers - 2	Nodes - 4	Accuracy - 15.15789
Layers - 2	Nodes - 5	Accuracy - 18.31579
Layers - 2	Nodes - 6	Accuracy - 16.63158
Layers - 2	Nodes - 7	Accuracy - 16.63158
Layers - 2	Nodes - 8	Accuracy - 13.05263
Layers - 2	Nodes - 9	Accuracy - 11.15789
Layers - 2	Nodes - 10	Accuracy - 14.73684
Layers - 3	Nodes - 1	Accuracy - 11.1579
Layers - 3	Nodes - 2	Accuracy - 13.26316
Layers - 3	Nodes - 3	Accuracy - 13.05263
Layers - 3	Nodes - 4	Accuracy - 10.10526
Layers - 3	Node

We can see that increasing the layers tends to become conterproductive and is better avoided. Also, the change in number of nodes doesn't affect the accuracy as significantly with lesser layers.

In [21]:
numoflayers = 1
numofnodes = 6

model=tf.keras.models.Sequential()
model.add(tf.keras.layers.Dense(xtr.shape[1],input_shape=(xtr.shape[1],), activation='relu'))
for i in range(1, numoflayers+1):
      nodes = xtr.shape[1]*numofnodes
      model.add(tf.keras.layers.Dense(nodes, activation='relu'))
      model.add(tf.keras.layers.Dropout(0.2))
model.add(tf.keras.layers.Dense(num_labels, activation='softmax'))

# Decide parameters
model.compile(optimizer=opt, loss=loss_fn, metrics=['accuracy'])
model.summary()

Model: "sequential_156"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_612 (Dense)            (None, 30)                930       
_________________________________________________________________
dense_613 (Dense)            (None, 180)               5580      
_________________________________________________________________
dropout_302 (Dropout)        (None, 180)               0         
_________________________________________________________________
dense_614 (Dense)            (None, 10)                1810      
Total params: 8,320
Trainable params: 8,320
Non-trainable params: 0
_________________________________________________________________


In [25]:
# Train model and get accuracy
model.fit(xtr, ytr, epochs=30, validation_data=(xv, yv))
model.evaluate(xv, yv)

Epoch 1/30
27/27 [==============================] - 0s 3ms/step - loss: 1.4966 - accuracy: 0.4184 - val_loss: 3.3049 - val_accuracy: 0.2105
Epoch 2/30
27/27 [==============================] - 0s 2ms/step - loss: 1.4814 - accuracy: 0.4314 - val_loss: 3.4594 - val_accuracy: 0.1895
Epoch 3/30
27/27 [==============================] - 0s 2ms/step - loss: 1.4805 - accuracy: 0.4350 - val_loss: 3.1891 - val_accuracy: 0.1684
Epoch 4/30
27/27 [==============================] - 0s 2ms/step - loss: 1.4199 - accuracy: 0.4539 - val_loss: 3.7009 - val_accuracy: 0.1579
Epoch 5/30
27/27 [==============================] - 0s 2ms/step - loss: 1.4267 - accuracy: 0.4444 - val_loss: 3.4979 - val_accuracy: 0.2105
Epoch 6/30
27/27 [==============================] - 0s 2ms/step - loss: 1.3873 - accuracy: 0.4645 - val_loss: 3.5533 - val_accuracy: 0.2105
Epoch 7/30
27/27 [==============================] - 0s 2ms/step - loss: 1.3966 - accuracy: 0.4669 - val_loss: 3.5694 - val_accuracy: 0.2000
Epoch 8/30
27/27 [==

[4.91787052154541, 0.2631579041481018]

It appears that we need a fairly simple model to get a good accuracy. The accuracy may not be "good" in a real world scenario but it is the best we get in the search space we explored.

In [26]:
real_accuracy=model.evaluate(xts,yts,verbose=0)
real_accuracy

[3.7647361755371094, 0.2666666805744171]

After tuning and using the best parameters,we can say that we get the accuracy of 26.67% has been achieved